cupy/core/core.pyx

```diff
@@ -1565,6 +1565,45 @@ cdef class ndarray:
     # TODO(okuta): Implement __setslice__
     # TODO(okuta): Implement __contains__
 
+    # numpy/ufunc compat
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        """
+        Apply unary or binary ufunc to this array
+
+        If binary, only allow if second argument is another cupy ndarray or
+        a number, i.e., raise ValueError instead of silently converting a
+        numpy array.
+        """
+        import cupy as cp  # top-level ufuncs
+        from numbers import Number
+        # pop because argument is already in inputs
+        if 'out' in kwargs:
+            # need to unfold tuple argument in kwargs
+            check = inputs + kwargs['out']
+            kwargs['out'] = kwargs['out'][0]
+        else:
+            check = inputs
+        for x in check:
+            if not isinstance(x, (ndarray, Number)):
+                # do *not* pass any old ndarray-like
+                return NotImplemented
```

Comment:
Passing to cupy results in
`TypeError: Unsupported type <class 'numpy.ndarray'>`
or using this as-is results in
`TypeError: operand type(s) all returned NotImplemented from __array_ufunc__`

Keeping the current means that, in theory, some other implementation could still work in combination with a cupy.ndarray, because NotImplemented implies trying the `__array_ufunc__` of the other object involved; but in practice, I think it can only realistically be cupy or a subclass. I thunk best would be to raise a TypeError with a carefully phrased message.

Reply:
My understanding is that CuPy could also return `NotImplemented` if they want to give up and hand to some other system and your solution here would just pass that through.  Is this correct?  If so then this probably isn't the right place to implement this logic.  Moreover, I don't think that you and I have enough experience with how CuPy is used to make policy decisions like this.  I think that we should avoid this discussion, or perhaps raise it as a separate issue that lives independently from this effort, which I think should be restricted to implementing the `__array_ufunc__` protocol.